### ISSUE TITLE
Support to async formula or variables

### ISSUE THREAD

**Issue by lucafaggianelli** [handsontable/formula-parser#31]
Hi,

is there the possibility to support async custom formulas or variables, for example something like this:
``` js
parser.setFunction('SUM_ASYNC', (params) => {
  return new Promise((resolve) => {
    return setTimeout(() => resolve(params[0] + params[1]), 1000))
  }
});

parser.parse('SUM_ASYNC(2, 3)').then((result, error) => { })
```

If you're wondering why, I would like to create formulas to access remote object, like `GET_FROM_DB('my_model', 123)`, could also be fine to have async variables.

Do you think is feasible? And how would you implement such a thing?

Thanks for your help!
Luca

**Comment by budnix** [comment]
Hi @lucafaggianelli,

Very interesting idea, I Like that! Looking through the code I see that this is not so trivial task as I suppose earlier. Currently, formula-parser strongly depends on https://github.com/zaach/jison as a formula grammar which doesn't support async values (I might be wrong). This should be more investigated.

**Comment by lucafaggianelli** [comment]
Thanks for your response! In the meantime, do you have any idea/workaround to solve the async formula problem?

**Comment by budnix** [comment]
I've investigated a little and found a workaround! Based on your snippet I've prepared working demo at https://www.webpackbin.com/bins/-KytJn6f7VgdYbSFApE5.

This simplified demo works but if you're using some built-in formulas such as `SUM` you will have to overwrite them to support Promises.

**Comment by lucafaggianelli** [comment]
wonderful! thanks a lot

**Comment by marek-baranowski** [comment]
@budnix webpackbin is down, could you move the code somewhere else, perhaps codesandbox? I'm also very interested in async functions.

**Comment by budnix** [comment]
I've tried recreating the demo and here are the results https://jsfiddle.net/uu6j4zax/1/. If you want to use async functions you have to overwrite all builtin functions to support Promises.

**Comment by aubrijesh** [comment]
Hi @lucafaggianelli,
I am looking for this features for some project where my function will resolve in async way. I went through this jsfiddle demo I found that this is really going to solve my issue. Do I need to override all the build in functions or I can go with only one async function define by me? Let us say I am writting a formula  "SUM(variable1, ASYNCFUN())". Now ASYNCFUN function will be resolved in async way. Please suggest me how can I implement this feature.

**Comment by budnix** [comment]
Unfortunately, you have to override all function that will support async/Promises (wrap original logic into `Promise.all` for example), as in the jsfiddle above.

**Comment by budnix** [comment]
And you can easily convert all formulas to async using this snippet:
```
// convert sync formulas to async function
formulaParser.SUPPORTED_FORMULAS.forEach((formulaName) => {
  parser.setFunction(formulaName, (args) => {
    return Promise.all(args).then((values) => {
      return formulajs[formulaName](...values);
    });
  });
});
```
like in this updated demo https://jsfiddle.net/uu6j4zax/20/.

**Comment by aubrijesh** [comment]
Thank you @budnix. Awesome work.

**Comment by aubrijesh** [comment]
Hi @budnix ,
I am trying to work with async solution and It is working perfectly with function but when I trying to give direct variable or static value it is failing.
Example :
 parser.parse('SUM(2, 5, C2)); 
is working fine as SUM and callCellValue are working in async way but when I am writting 
1- parser.parse('CONCATENATE(SUM(2, 5, C2), name, age, "hello world")');
2- parser.parse('SUM(2, 5, C2) + 12+ age');
is not working even though I have return promise in callVariable.
I have created fiddle for this. Please have a look on this. 
https://jsfiddle.net/brjiesh/5tgsbd7y/4/

**Comment by aubrijesh** [comment]
Hi @budnix,
Thank you so much for providing solution in such a short time. You have saved my life @budnix, Thank you again.

**Comment by pbadenski** [comment]
Just FYI - be careful when applying this approach. Some functionality I believe cannot be properly adapted (eg. expression such as "-A1"), because:
```
var n1 = yy.invertNumber($$[$0]);

this.$ = n1;
// n1 is a Promise, can't do anything to stop the code below from triggering :(
if (isNaN(this.$)) {
   this.$ = 0;
}
```

**Comment by vishal767** [comment]
yep .. nested async functions fails `SUM(1+FROM_TASK('ZcThzAVBhD','true','true'))` ,FROM_TASK is async , and this fails
